### ISSUE TITLE
QtSLiM needs some kind of installation plan

### ISSUE THREAD

**Issue by bhaller** [MesserLab/SLiM#78]
Might be of interest to @philippmesser and @petrelharp and @brnorris03 and @samchamper and @grahamgower, feel free to tag others.

QtSLiM, the Linux port of SLiMgui, is making good progress and will be ready to ship in perhaps the next month-ish.  Most things now work, usually more or less exactly as they work in SLiMgui; I'm just implementing a handful of remaining features.  I'm figuring to roll it out as part of a SLiM 3.4 general release.  (At present it's in the qtslim branch of the SLiM repo; feel free to check it out, you'll need to install Qt first.)

However, there are a number of issues that I don't really know how to deal with, partly because of my cluelessness about Linux, partly because of my cluelessness about makefiles and Unix-style builds/installs.  I'm not really even sure of the right overall strategy on many of these.  So I'm creating an issue where hopefully discussion can occur, and maybe solutions can be generated.

(1) What is the overall installation strategy on Linux?  Should we just tell people to build QtSLiM from sources?  Or should there be some kind of solution involving apt-get or whatever, or a double-click installer?  I would have no idea how to do those latter things.  Do Linux people care?  My impression that (unlike Mac people) they don't mind building from sources and are used to it.  Is that generally true?

(2) What requirements should we state?  I've been aiming for Qt 5.9.8 as a minimum, but it presently appears that some things might be easier if we required a more recent version of Qt; does anybody care about this, or can we require basically the current version of Qt if we want to?  Similarly: I have only been testing on Ubuntu 18.04 LTS.  Is it OK to state that as a requirement, and ignore bugs related to other distros, other Ubuntu versions, etc.?  Otherwise it's hard to see what to do; I don't feel like I can possibly try to handle installation and testing on all the different Linux platforms people might want to use, that would require a dedicated QA engineer.

(3) How do we associate an icon with QtSLiM, and how do we assign it responsibility for the ".slim" file extension, on Linux?  Related to the former, I have found this: https://askubuntu.com/questions/894990/how-to-change-an-icon-in-16-04.  Related to the latter, I have found this: https://askubuntu.com/questions/538299/globally-associate-file-type-with-certain-application.  Both seem to involve modifying a sort of application registry file, at least on Ubuntu.  I don't know how general this is to other distros, or even other windowing systems, nor do I really understand what is involved.  It seems like it would be a good idea to have a script we provide that does the necessary system modifications for the user; and perhaps that script could even be automatically run at the end of a "make install".  Does that sound right?  Anybody know how to actually do this?

(4) Regarding building from sources, right now we have two parallel ways to do it: with qmake/make (qmake being a make tool built into Qt), or with cmake/make (cmake being the standard make strategy we already use for everything else with SLiM).  Since I use Qt Creator for development of QtSLiM, the qmake/make route is a part of my standard development workflow and will necessarily continue to be maintained.  Does it make sense to *also* supply/support the cmake-based strategy, which does presently seem to work on Linux on Travis-CI (thanks Boyana), or should we just declare that QtSLiM is built with qmake/make, period?  Are there advantages to cmake over qmake here?

(5) A big question is: what does all this imply for SLiMgui?  QtSLiM has turned out quite well, I think, and feels quite platform-native on macOS, although there are a few places where it's less than fully native.  Maintaining two separate but parallel codebases for two GUI apps is a lot of added work, for little advantage.  I'm therefore pondering the idea of retiring SLiMgui once QtSLiM is done.  Maybe that would mean that QtSLiM gets renamed "SLiMgui" and inherits its place.  And maybe that means I should actually call QtSLiM "SLiMgui" now, when I release it on Linux, to avoid confusing people; maybe there's no reason for it to have a different name.  I'm up in the air on all of these questions.

Thoughts?

**Comment by samchamper** [comment]
I can't speak for all Linux SLiM users, but for my part:

1. I think getting SLiM on a package manager could probably be more of a stretch goal. As convenient as apt-get is, my default assumption is that it won't have the latest version of any software that I'm interested in installing. And I'm already used to installing SLiM from source.

2. I think, so long as you have things fully tested and functioning on one distro, it's okay to leave it to folks who are using other distros to figure out how to get things to that state. And I think Ubuntu 18.04 LTS is a very reasonable choice. As for the version of qmake - I myself don't use it to build any other software, so I don't have any reason not to just upgrade to the latest version. I suppose installing an up to date version of qmake might be a slight annoyance in a shared computing environment, but certainly not to the extent that you should put more work on yourself to avoid that. As far as I'm concerned, there's no reason not to use whatever makes life easiest for you.

3. No clue about this!

4. Hmm... I guess I could imagine that some shared or cloud computing environments have cmake but installing qmake might be a bit of a hassle? And one wouldn't care about the GUI tools in such an environment anyway. For such cases, keeping the cmake workflow would be handy. On the other hand, if maintaining both workflows creates a lot more work for you, I suppose that the small advantage of having both might not be worth it.

5. If you really think the Qt version sacrifices nothing, unifying the code base seems like a great call. On the other hand, you could wait until the Qt version gets a little bit of user-testing first, and then unify the code base after the new GUI has proved itself!

**Comment by petrelharp** [comment]
1. Agree with @samchamper. Distributing a Debian package is not too hard but just listing prerequisite packages and giving build instructions is almost as good. We can just start with the distros we have and others can contribute instructions for others.

2. Also agree. I think that LTS Ubuntu releases are a least common denominator, but more recent Qt could be required if there's a good reason! Folks can open issues for other distros if necessary.

3. I can have a go at figuring this out.

4. The cmake setup seems to work now for me.

5. I vote for only one way to do a thing. Maintaining multiple implementations sounds terrible.

And, wow, I am really impressed. This is wonderful.

**Comment by grahamgower** [comment]
1. If you have resources to build packages, that's fantastic. But this certainly shouldn't be a priority. Because of the diversity of linux distributions and packaging options, you'll certainly need to rely on volunteers.
2. I don't know Qt nor Ubuntu. Looking at the Qt git repository though, it looks like v5.9 was tagged three years ago, and they tagged a bugfix release v5.9.9 in December last year. So v5.9.something sounds like a very reasonable requirement.
3. Not sure.
4. I also tried the `cmake` build, which worked fine. My suspicion is that this will be the easier way for people to build. I've seen cmake before and can read the man page to find that `CMakeLists.txt` is the place for fine-tuning. In contrast, I've never invoked `qmake` before. I seem to have it installed, but `man qmake` says that no manual page exists. So even figuring out how to call qmake would need more time invested.
5. QtSLiM FTW!

PS: I was about to volunteer to make an `Arch Linux` package for SLiM. But then I noticed one already exists! :P
https://wiki.archlinux.org/index.php/SLiM

**Comment by grahamgower** [comment]
4. Addendum... It seems one just runs `qmake` in the top-level directory, which creates a regular `Makefile`. But after a minute or so, it threw a compiler error. :)

**Comment by brnorris03** [comment]
> (1) What is the overall installation strategy on Linux? Should we just tell people to build QtSLiM from sources? Or should there be some kind of solution involving apt-get or whatever, or a double-click installer? I would have no idea how to do those latter things. Do Linux people care? My impression that (unlike Mac people) they don't mind building from sources and are used to it. Is that generally true?

It's nice to have a Debian package (works on many distros), but of course, someone has to maintain it, i.e., update for new distro releases in addition to the SLiM updates. Those are not that frequent, so perhaps this is fairly easy.

> (2) What requirements should we state? I've been aiming for Qt 5.9.8 as a minimum, but it presently appears that some things might be easier if we required a more recent version of Qt; does anybody care about this, or can we require basically the current version of Qt if we want to? Similarly: I have only been testing on Ubuntu 18.04 LTS. Is it OK to state that as a requirement, and ignore bugs related to other distros, other Ubuntu versions, etc.? Otherwise it's hard to see what to do; I don't feel like I can possibly try to handle installation and testing on all the different Linux platforms people might want to use, that would require a dedicated QA engineer.

I think the Qt and compiler versions are the only things that really matter, I'd pick the lowest that works as the requirement. This is for CMake builds only, obviously, a package would work with whatever version the Linux distro has.

> (3) How do we associate an icon with QtSLiM, and how do we assign it responsibility for the ".slim" file extension, on Linux? Related to the former, I have found this: https://askubuntu.com/questions/894990/how-to-change-an-icon-in-16-04. Related to the latter, I have found this: https://askubuntu.com/questions/538299/globally-associate-file-type-with-certain-application. Both seem to involve modifying a sort of application registry file, at least on Ubuntu. I don't know how general this is to other distros, or even other windowing systems, nor do I really understand what is involved. It seems like it would be a good idea to have a script we provide that does the necessary system modifications for the user; and perhaps that script could even be automatically run at the end of a "make install". Does that sound right? Anybody know how to actually do this?

I don't know that Linux people care much about relying on extensions in general, but instead of global modifications, usually you would have "#!/path/to/executable" in the first line of your executable script, in this case slim.

> (4) Regarding building from sources, right now we have two parallel ways to do it: with qmake/make (qmake being a make tool built into Qt), or with cmake/make (cmake being the standard make strategy we already use for everything else with SLiM). Since I use Qt Creator for development of QtSLiM, the qmake/make route is a part of my standard development workflow and will necessarily continue to be maintained. Does it make sense to _also_ supply/support the cmake-based strategy, which does presently seem to work on Linux on Travis-CI (thanks Boyana), or should we just declare that QtSLiM is built with qmake/make, period? Are there advantages to cmake over qmake here?

CMake was meant for building SLiM in the absence of Qt -- this is our typical scenario. Some HPC serves do not (and never will) have Qt installed.

> (5) A big question is: what does all this imply for SLiMgui? QtSLiM has turned out quite well, I think, and feels quite platform-native on macOS, although there are a few places where it's less than fully native. Maintaining two separate but parallel codebases for two GUI apps is a lot of added work, for little advantage. I'm therefore pondering the idea of retiring SLiMgui once QtSLiM is done. Maybe that would mean that QtSLiM gets renamed "SLiMgui" and inherits its place. And maybe that means I should actually call QtSLiM "SLiMgui" now, when I release it on Linux, to avoid confusing people; maybe there's no reason for it to have a different name. I'm up in the air on all of these questions.

Maybe before retiring, make QtSLiM the default GUI and collect some feedback from previous users?

**Comment by bhaller** [comment]
Hi all.  Thanks for all your input, this has been helpful.  BTW, if you have QtSLiM bugs to report please feel free to create separate issues on GitHub for them.

OK, summing up and reacting to each of the original points (1)-(5):

(1) Installation strategy.  Seems to be a consensus that I should not worry about any kind of "installer" for Linux, and just let people build from sources; maybe third parties will set up installers for particular platforms (@brnorris03 suggests Debian in particular?), but this is not my problem.  Good.

(2) System requirements.  Seems to be a consensus that stating that it has been tested only on Ubuntu 18.04 LTS, and letting users of other distros file bugs/PRs if there are problems, is acceptable.  Good.  Also consensus that being compatible with Qt 5.9.8. and later would be ideal, but that that requirement can move to the current Qt (5.14.2) if necessary.  (Not yet clear whether it is necessary.)

(3) App icon, app .slim extension registration.  Nobody has a magic wand for this – it's weird to me that Qt itself doesn't make this easy! – but @petrelharp said he'd look into writing a script users could run after make (or perhaps as a make install type of target?), which would be fantastic.  Re: @brnorris03's comment about shebangs ("#!/path/to/executable") to associate a script with a tool: I could imagine supporting that for the command-line slim tool (although right now trying to do it would produce a syntax error in slim; nobody has ever asked for that to work), but that seems like a weird idea to try to associate a script with a GUI executable like QtSLiM, no?  Even in the Linux world it's normal for specific file extensions like .pdf, .jpg, .txt, etc., to be associated with a default GUI app, right?  That's what I'm talking about – making it so double-clicking a .slim file opens the file in QtSLiM.

(4) Building with qmake/make vs. cmake/make.  I think I confused people with this one.  I was not suggesting retiring cmake building for SLiM in general; that is obviously the right path forward and works well.  I was just asking specifically about build QtSLiM, which is the one build target that relies upon Qt anyway, and perhaps it is natural to just use qmake for it (since that has to be set up anyway for me to use Qt Creator as a development environment when I'm working on QtSLiM).  Discussion on this point seems inconclusive so far.  All my local testing is pretty much done through Qt Creator, and thus through qmake; the Travis-CI builds use cmake, though.  So both are getting their builds validated, at least, but only the qmake build is actually getting tested by me, at present.

(5) SLiMgui vs. QtSLiM, going forward.  Seems to be consensus that QtSLiM can inherit the kingdom if it ends being just as complete/useful/stable as SLiMgui is, and that maintaining two parallel GUI codebases forever would be crazy.  Agreed.  Unclear how the transition will be done; i.e., whether QtSLiM will eventually get renamed "SLiMgui", or what.  But there's a general feeling that it would be wise to initially release QtSLiM as a new app, parallel to SLiMgui, rather than trying to transition over right at the point of initial release, since there might be big problems with QtSLiM that aren't found prior to initial release.  I agree.

All this is quite clarifying.  Thanks!

P.S. @grahamgower, that Arch Linux package for SLiM that you found is for something else named Slim, not for SLiM.  :->

**Comment by grahamgower** [comment]
> P.S. @grahamgower, that Arch Linux package for SLiM that you found is for something else named Slim, not for SLiM. :->

It's kinda interesting that they even used the same capitalisation. :) I've got an Arch Linux package ready to go now, but unfortunately it will have to get a different name (I called it `slim-simulator` for now).

**Comment by petrelharp** [comment]
> I've got an Arch Linux package ready to go now

Wow, that was quick!

I once asked a student to install SLiM and they spent quite a while trying to install the [slim php framework](http://www.slimframework.com)...

**Comment by bhaller** [comment]
Hi everyone.  SLiM 3.4, with QtSLiM, is just about ready to ship; I'm aiming for early next week.  I'd appreciate it if folks here could give it a quick test, building it from scratch.  It is now on the master branch of the SLiM repository; the qtslim branch has been merged in and deleted (finally).  The new manual, available at http://benhaller.com/slim/SLiM_Manual_3.4.pdf for now, has detailed instructions for building QtSLiM (section 2.2.2) and for configuring the GNOME desktop environment on Ubuntu to recognize the app and the .slim file extension (section 2.2.3); it would be great if folks could give those sections a read-through and send comments/changes based on their particular Linux platforms.  Note that the executable produced by cmake is now named `QtSLiM`, not `qtslim` (the lowercase name was unintentional).  Be sure to clean your old build folder completely (I recommend just deleting it and making a new one) so that nothing from old builds carry over; cmake and qmake are not always great about cleaning up their own mess, and it can cause bizarre compile/link problems.

Thanks for any feedback!  This issue is a good place for any chat; I plan to close it once SLiM 3.4 ships, since the questions it raises have been resolved.

**Comment by bhaller** [comment]
@grahamgower @petrelharp @brnorris03 @samchamper @vsbuffalo @philippmesser

**Comment by bhaller** [comment]
If anybody tagged here plans to look at this beta release in the next couple of days, please reply.  Otherwise I guess I'll just release it without external testing; but that is not optimal, so if you can give it a try, that would be great.  I'm actually even more interested in comments on the QtSLiM building instructions in the manual than I am in testing of QtSLiM itself.  Thanks all.

**Comment by brnorris03** [comment]
Finally done with meetings, I can give it a whirl today!
On 5/5/2020 1:55:21 PM, Ben Haller <notifications@github.com> wrote:
If anybody tagged here plans to look at this beta release in the next couple of days, please reply. Otherwise I guess I'll just release it without external testing; but that is not optimal, so if you can give it a try, that would be great. I'm actually even more interested in comments on the QtSLiM building instructions in the manual than I am in testing of QtSLiM itself. Thanks all.

**Comment by brnorris03** [comment]
I tested on OSX, Catalina, using mac ports; the only slightly tricky bit was that qmake wasn't automatically in my path after running `sudo port install qt5`, so I had to either change the path or do `sudo ln -s /opt/local/libexec/qt5/bin/qmake /opt/local/bin/qmake`. 
```
$ qmake -v
QMake version 3.1
Using Qt version 5.14.2 in /opt/local/libexec/qt5/lib
```
After that, I followed the instructions on p. 54 of the manual, and everything built. There were many warnings about deprecated Qt functions, so in future these would probably have to be addressed. 

The only runtime issue I encountered has to do with the color scheme -- it picked up some of the colors from the dark theme used in the system settings, but not all, resulting in less than ideal contrast for buttons and such. I imagine this can happen in Linux environments, too, although I haven't tested there yet (with a dark theme).

<img width="1062" alt="Screen Shot 2020-05-05 at 17 27 05" src="https://user-images.githubusercontent.com/3604514/81128487-33c6ff80-8ef6-11ea-8a2a-6f0a3e7a8b72.png">
<img width="812" alt="Screen Shot 2020-05-05 at 17 27 25" src="https://user-images.githubusercontent.com/3604514/81128500-3b86a400-8ef6-11ea-9778-f48ea25a50e6.png">

**Comment by bhaller** [comment]
Very interesting, great feedback @brnorris03.  Incidentally, the deprecated API warnings are known/irrelevant; they are disabled for the qmake build, and are because I target Qt 5.9.8 as the minimum version.  I'll see if I can figure out how to disable them for the cmake build.  I'll work on the other things, good stuff!

**Comment by brnorris03** [comment]
On my Ubuntu laptop (which has a little screen, but very high resolution). The colors look great, however, the fonts in the widgets are very small, the actual icons are also tiny, bu I can still kind of see them. The main problem is that some text is obscured, see screenshot. Things don't improve with resizing because the individual subsections are not resizable. 
![Screenshot from 2020-05-05 17-59-09](https://user-images.githubusercontent.com/3604514/81129871-a4701b00-8efa-11ea-90cc-bdd81e7e5bb0.png)

**Comment by bhaller** [comment]
Hi @brnorris03.  Wow, that looks awful.  It's very interesting that the text in the status bar at the bottom of the window is in tiny type as well; that's a widget that is defined and set up by Qt, not by me.  So that seems to indicate that the screw-up here in in Qt itself; it is probably confused by the high-res screen of your laptop.  What version of Qt are you building against on that machine?  If it is the default for Ubuntu 18.04 LTS, which is Qt 5.9.5, can you install Qt 5.14.2 or some similar recent version (the new SLiM manual has instructions) and see if that fixes the problem?  I believe they have made many fixes related to high-DPI screens since 5.9.5.

I would also be interested to know whether there is a value set for either of the `QT_SCALE_FACTOR` or `QT_FONT_DPI` environment variables on your machine, and whether setting a new value for them helps.  For example, launching QtSLiM with `QT_SCALE_FACTOR=1.3 QtSLiM` or `QT_FONT_DPI=128 QtSLiM` might be interesting.

Unfortunately I can't reproduce this problem on my end; I've got two machines here with high-DPI screens, but QtSLiM seems to work well on both.  So any further info would be useful.

**Comment by brnorris03** [comment]
Okay, installing 5.14 now (it will take a while, my network is slow). In the meanwhile I tried scaling the 5.9 default version with the env. variables, and it does improve things (but couldn't get it to both show all text and be large enough). Will update for 5.14 when that finishes installing.

![Screenshot from 2020-05-05 18-40-16](https://user-images.githubusercontent.com/3604514/81131470-1860f200-8f00-11ea-92fb-946fe76e1817.png)

**Comment by bhaller** [comment]
Hi @brnorris03.  Well, that does look considerably better.  Looks like only the "Generation" label is clipped now?  I think the layout is assuming that the widths of the buttons above it suffice to guarantee sufficient room for the label text.  I'll look into fixing that assumption.  Thanks!

**Comment by petrelharp** [comment]
I'm trying it out now!

**Comment by petrelharp** [comment]
It works! I'm on
```
$ qmake --version
QMake version 3.1
Using Qt version 5.12.5 in /usr/lib/x86_64-linux-gnu
```

Minor suggestions:

I think you need a line break between the sentences here:
```
I am now using Qt 5.14.2 on my Ubuntu machine. On macOS, the simplest solution seems to
```

You may want to suggest making a different `build/` directory for qtslim? You did say previously that this is possible, but by changing the example code (`mkdir qtbuild`...) that'd get more people to do it, and not get messed up.  Or - does setting `BUILD_QTSLIM=ON` affect `slim` and `eidos`? If not, then no big deal, I guess?  If so, you should maybe say so (where you say "if you want to build slim and eidos as well, you can simply do make .").

No errors about deprecated Qt functions. Everything looks reasonable:
![Screenshot from 2020-05-05 21-42-44](https://user-images.githubusercontent.com/1046249/81139316-291e6180-8f1a-11ea-8991-3cf08ba49424.png)

I don't have a dark mode handy.

There *is* a bug, though: alt-clicking to bring up the help doesn't recognize when the script window has scrolled, so it brings up the help for whatever would have been under my cursor if I was still at the top of the script.  I'll try upgrading to Qt 5.14 and see if that clears it up.

**Comment by petrelharp** [comment]
Also, omg about the current Qt situation.

**Comment by grahamgower** [comment]
Sorry for the late response @bhaller. Everything still works for me with the master branch. I read through the new manual bits, and they seem fine (although I don't have Ubuntu nor GNOME, so I didn't test the instructions). I've updated the Arch Linux package to use the master branch, and will update it again whenever you tag new releases: https://aur.archlinux.org/packages/slim-simulator/

The alt-click bug @petrelharp mentioned above is reproducible with Qt 5.14.2.

**Comment by petrelharp** [comment]
So: I have installed Qt 5.14.2 from the company's installer, following your directions (it took a LONG time, and I had to make a Qt account...).  Now, doing `qmake --version` gives me `5.14.2`.

However, based on `lsof -p QtSLiM` it doesn't look like QtSLiM is actually using 5.14.2, even after making the qtchooser file, and also after setting the `Qt5_DIR` environment variable as you suggest. I haven't been able to figure out how to tell if I'm using the right version, even?

Other notes:

- From the man page for `qtchooser` it looks like putting the default configuration in `~/.config/qtchooser/default.conf` would do the same thing as `/etc/xdg/qtchooser/default.conf`, and not require all the sudo- and admin:-ing.

- It is much easier for people to install Qt from their package manager; so perhaps you should say that "Installing your distribution's Qt5 package may get you a sufficiently recent version: in Debian/Ubuntu you should install qt5-default, and in Arch the package is called qt5-base."  The Qt version in Debian unstable is [5.12.5](https://packages.debian.org/sid/qt5-default); in Debian stable it is [5.11.3](https://packages.debian.org/buster/qt5-default); and it might even be that installing from experimental would get you [5.14.2](https://packages.debian.org/experimental/qt5-default) and not break things.

- The directory in
```
Qt5_DIR=/Users/bhaller/Qt5.14.2/5.14.2/clang_64 cmake ...
```
isn't consistent with your previous example directories.

**Comment by petrelharp** [comment]
Another note: I had to add a `-t` here:
```
sudo gtk-update-icon-cache /usr/share/icons/Humanity -f -t
```
or else I got an error:
```
gtk-update-icon-cache: No theme index file.
```
... rebooting now to see if it works.

**Comment by petrelharp** [comment]
And, It works!!! :tada:  At least, `xdg-open something.slim` opens up QtSLiM, and I have a SLiM icon in the sidebar! I'm not using a desktop, so I can't check this double-clicking thing, but I believe it works via `xdg-open`.

**Comment by brnorris03** [comment]
Yes, just Generation is clipped, it's not a big deal. The icons are a bit small, but I haven't tinkered too much with the scaling; I can still read them. I did have to set both QT_SCALE_FACTOR and QT_FONT_DPI. It also looked exactly the same with 5.14.2 (still needs those variables).

**Comment by brnorris03** [comment]
> Hi @brnorris03. Well, that does look considerably better. Looks like only the "Generation" label is clipped now? I think the layout is assuming that the widths of the buttons above it suffice to guarantee sufficient room for the label text. I'll look into fixing that assumption. Thanks!

Yes, just Generation is clipped, it's not a big deal. The icons are a bit small, but I haven't tinkered too much with the scaling; I can still read them. I did have to set both QT_SCALE_FACTOR and QT_FONT_DPI. It also looked exactly the same with 5.14.2 (still needs those variables).

**Comment by bhaller** [comment]
Great feedback everybody, thanks!  I'll just reply to everything here, rather than replying to each message separately.  Note that I have not updated the draft manual PDF at this time; I'm still doing edits.  I think there are action items for each of you below, if you're willing; so please read on.

- Re: `qmake` not being in the path on macOS (@brnorris03), yes; I guess that's unclear, but since Qt isn't preinstalled on macOS anyway I didn't expect Mac users to even try to run that command.  :->  I've added a bit of discussion of that path issue, and suggested symlinking as you said.  FIXED.

- Re: deprecated function warnings (@brnorris03), I'll look into suppressing them when building with `cmake`.

- Re: dark mode on macOS (@brnorris03), I have just committed a fix that forces QtSLiM to run in light mode regardless of the system setting.  FIXED.  (In the longer term it would be nice to actually support dark mode properly, but that's for next release.)

- Re: the clipped text and scaling issues (@brnorris03), I think you might want to use `QT_SCALE_FACTOR=2.0` instead of `1.5`.  The buttons in the UI in your screenshot are still smaller than they're supposed to be, and that means that not enough room is given in the layout for the full "Generation" label to display (it also causes some other small appearance glitches).  So, see what you think with `QT_SCALE_FACTOR=2.0`.  I suspect `QT_FONT_DPI=144` is also the correct value for your situation, since the default is probably 72 pixels per inch, and you want to double it.  I'm going to specifically mention in the manual that users on high-DPI screens may want to try these settings.  I'm not sure that there's anything I can do about this problem in my code – I think Qt is failing to detect that your laptop screen is a high-DPI display – but anyway these settings should fix the problem for users, I think.  FIXED (I guess).

- Re: the build directory (@petrelharp), `BUILD_QTSLIM=ON` has no effect on `slim` and `eidos`; they will build as usual.  I've added a bit of text to clarify that.  FIXED.

- Re: alt-clicking not working properly when the script view is scrolled (@petrelharp), good catch.  I was missing a coordinate system transform.  FIXED.

- Re: the Arch Linux package (@grahamgower), great.  Should I mention the existence of this package in the manual?  And if so, what exactly should I say?  Something like "Users on Arch Linux can install QtSLiM by executing the package installation command <...>", I suppose?  Can you fill in the blank for me?

- Re: which version of Qt ends up actually getting built against and then executed against (which are not necessarily the same)... hmm (@petrelharp).  For now, (a) I have added display of the Qt version in QtSLiM's about box so one can easily check the version one is running against, and (b) I have added code that will emit a console log and exit if QtSLiM detects that it is running against a different version of Qt than it was built against, so now the two ought to always match.  But you might still not be building against the Qt version you intend, and I'm surprised that using `Qt5_DIR` didn't help with that.  Grr.  This stuff is so complicated, and so poorly documented.  I'll see if I can figure out more about that.  I think one problem *might* be that the QtSLiM build is linking to Qt dynamically, not statically, but I'm not sure that's true, and I'm not sure how to fix it.  @brnorris03 do you have input on this?  How I would tell `cmake` to static-link against Qt rather than dynamic-link?  Well, I'll bang my head against that problem tomorrow.

- Re: `qtchooser` and `~/.config` instead of `/etc` (@petrelharp), yes, it sounds like that *should* work, but I saw people online saying that it does not, in fact, work on some distros; people recommended using `/etc`.  In any case, `/etc` sets it up for all user accounts on the machine, which is (arguably) desirable.  FIXED (not a bug, I think).

- Re: installing for a package manager (@petrelharp).  Well, I couldn't get it to work on Ubuntu, but that was a while ago, before I had figured some of these things out.  I'll try it again; if it works smoothly it would provide a much cleaner story than all this crap.  Peter, can you flesh out your proposed sentence that starts with "Installing your distribution's..."?  Specifically: what exact command line would be used to do those things, on Debian/Ubuntu and on Arch, and how would one choose between "unstable", "stable", and "experimental" (i.e. how does one modify the install command to make that choice), and how would one find out which Qt version is associated with each choice at the present time?

- Re: the `Qt5_DIR` path being inconsistent with previous doc (@petrelharp): Yes, that's because that bit is discussing macOS, not Linux.  I'm starting to think I need to split this section of the manual in two, with one section for Linux and one for macOS; discussing them in an interleaved fashion is clearly confusing.  I wasn't really even intending to get into building on macOS, for this release, but I seem to have ended up there.

- Re: `-t` for `gtk-update-icon-cache` (@petrelharp): OK, added this to the procedure.  I thought that index would always be present for `/usr/share/icons/Humanity`, but apparently not.  All this icon cache stuff seems to be semi-broken; it's a miracle it works at all.  FIXED.

OK, so, that's where it's at for now.  Progress on some items, some unresolved.  Thanks folks!

**Comment by bhaller** [comment]
Figured out the deprecated headers issue, which wasn't what I thought it was; it's due to using older OpenGL APIs.  That causes deprecation warnings on macOS, which I have now suppressed.  (I'm not going to get off of those old APIs unless/until I actually have to.)  FIXED.

**Comment by petrelharp** [comment]
>  Peter, can you flesh out your proposed sentence that starts with "Installing your distribution's..."? 

"
Installing your distribution's Qt5 package may get you a sufficiently recent version: in Debian/Ubuntu, you can do this by running
```
sudo apt-get install qt5-default
```
In Arch, run
```
pacman -S qt5-base
```
"

(I am not certain about the Arch command.  Also: it's amazingly great that @grahamgower has got a package together, and you should definately mention it.)

As for the stable/unstable/experimental thing - if for instance you wanted people to get Qt5.14 then you could point them towards the experimental package (and I could assist); but it's really best to just leave it at the above, if you're happy with 5.11.3.  (And, if so, maybe the sentence above should be "You can install Qt5 on Debian/Ubuntu by running..." instead of the "may get you..." bit.)

**Comment by petrelharp** [comment]
As for which version of Qt is being used: 
```
$ lsof -p 19235 | grep libQt
QtSLiM  19235 peter  mem       REG                8,1   366280  398125 /usr/lib/x86_64-linux-gnu/libQt5Svg.so.5.12.5
QtSLiM  19235 peter  mem       REG                8,1   619952  411013 /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5.12.5
QtSLiM  19235 peter  mem       REG                8,1  1609928  399702 /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5.12.5
QtSLiM  19235 peter  mem       REG                8,1  5458184  452888 /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.12.5
QtSLiM  19235 peter  mem       REG                8,1  6116736  399700 /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.12.5
QtSLiM  19235 peter  mem       REG                8,1  6831816  399704 /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.12.5
```
... so it's definately using the system's shared libraries, dynamically. It ran (and I didn't see anything in the console log), so I guess that's the one I built against anyhow, although I don't see a version in the About QtSLiM box either.

**Comment by petrelharp** [comment]
>  I'm starting to think I need to split this section of the manual in two, with one section for Linux and one for macOS

Yes, I think that's wise.

**Comment by petrelharp** [comment]
> I'm surprised that using Qt5_DIR didn't help with that. Grr.

Ack, sorry!!! I have done this again, but this time in not such a tired state, and in fact this:
```
Qt5_DIR=/home/peter/lib/Qt/5.14.2/gcc_64 cmake -D BUILD_QTSLIM=ON ..
make QtSLiM
```
*does* compile with Qt 5.14.2. I can't guess what I did wrong before.  And, running it is using the correct shared libraries:
```
$ lsof -p 22414 | grep libQt
QtSLiM  22414 peter  mem       REG                8,3   447816 6612981 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5Svg.so.5.14.2
QtSLiM  22414 peter  mem       REG                8,3   756888 6611681 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5DBus.so.5.14.2
QtSLiM  22414 peter  mem       REG                8,3  2006008 6560379 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5XcbQpa.so.5.14.2
QtSLiM  22414 peter  mem       REG                8,3  6945448 6611680 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5Core.so.5.14.2
QtSLiM  22414 peter  mem       REG                8,3  8668192 6611684 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5Gui.so.5.14.2
QtSLiM  22414 peter  mem       REG                8,3  8085232 6560375 /home/peter/lib/Qt/5.14.2/gcc_64/lib/libQt5Widgets.so.5.14.2
```
I don't see any differences between the two - on the left is the 5.12.5 version and on the right is 5.14.2:
![Screenshot from 2020-05-06 21-33-40](https://user-images.githubusercontent.com/1046249/81256115-5afbfa80-8fe4-11ea-976e-ed5f679c62a7.png)

To compile using my installed 5.14.2, `Qt5_DIR` does seem to be necessary - I tried the various permutations of setting the qtchooser conf again, and none of them worked. Is there anything you want me to try out with 5.12 versus 5.14? I'd much rather use my debian package Qt unless there's a good reason to use 5.14.

**Comment by grahamgower** [comment]
> * Re: the Arch Linux package (@grahamgower), great.  Should I mention the existence of this package in the manual?  And if so, what exactly should I say?  Something like "Users on Arch Linux can install QtSLiM by executing the package installation command <...>", I suppose?  Can you fill in the blank for me?

It would be great to have it mentioned, but detailed instructions shouldn't be necessary (Arch users are typically familiar with packaging details, or they wouldn't even get the OS installed). Maybe something like: "Acquire the PKGBUILD file from https://aur.archlinux.org/packages/slim-simulator and install with `makepkg -si`. More detailed instructions for installing packages from the Arch User Repository can be found on the Arch wiki https://wiki.archlinux.org/index.php/Arch_User_Repository#Installing_packages"

@petrelharp is indeed correct that the command to install qt is `pacman -S qt5-base`. This shouldn't be necessary if the user passes the `-s` flag to `makepkg` when installing, as the dependencies will be automatically pulled in.

**Comment by brnorris03** [comment]
>  @brnorris03 do you have input on this? How I would tell cmake to static-link against Qt rather than dynamic-link? Well, I'll bang my head against that problem tomorrow.

I don't think you have to do that, it will store the path to the dynamic library in the executable. It would be nice for CMake to print out the one it ended up using, though, which is just a matter of adding a message when it's found. For any executable, you can check the paths and libraries with `ldd <exe>` on Linux or `otool -L <exe` on OSX.  On my mac:
```
$ otool -L ./QtSLiM 
./QtSLiM:
	/opt/local/libexec/qt5/lib/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.14.0, current version 5.14.2)
	/opt/local/libexec/qt5/lib/QtGui.framework/Versions/5/QtGui (compatibility version 5.14.0, current version 5.14.2)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/opt/local/libexec/qt5/lib/QtCore.framework/Versions/5/QtCore (compatibility version 5.14.0, current version 5.14.2)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

**Comment by bhaller** [comment]
In the long run I'd like to static-link to Qt, because that way I'll be able to distribute a binary for macOS that will be self-contained, just like the current double-click installer.  I gather this is pretty standard for Qt apps, as a way to make the distribution of the app straightforward; you don't really want to make end users build and install the Qt framework.

**Comment by brnorris03** [comment]
It's not difficult to make cmake require static libraries, but it would have to be tested on various platforms since not all packages include static libraries. You would be adding -static in all the target_link_library directives, e.g., 
`target_link_libraries( ${TARGET_NAME} PUBLIC Qt5::Widgets Qt5::Core Qt5::Gui OpenGL::GL gsl tables eidos_zlib -static )`

This would fail if there are no static libraries available, of course. So on Ubuntu, for example, libOpenGL is only available as shared library by default.

**Comment by bhaller** [comment]
I see.  Can one link dynamically against some libs and statically against others?

**Comment by brnorris03** [comment]
Yes, in general, but from what I understand about Qt and its dependencies, it is not easy or possible with normal distros and may require a build from source.

**Comment by bhaller** [comment]
I see.  Hmm.

**Comment by bhaller** [comment]
Hi all.  I've completely reworked the QtSLiM install doc; I split the macOS stuff into a separate subsection, and I switched from the Qt5 "Online Installer" as the focus of the procedure to using the apt-get installer as the focus of the procedure, since it seems to work much more smoothly.  I've put up a new draft of the manual at http://benhaller.com/slim/SLiM_Manual_3.4.pdf.  I'm going to spend a couple of hours today running through the procedures one last time to make sure there are no typos etc.  If any of you have time to try this out once more, that would be great; but I know you've all spent a lot of time on it already, so no worries if you're busy.

I've also worked out how to make a build of QtSLiM on macOS that can be installed on its own, without the need for the end user to install Qt at all.  I plan to distribute such a build as a separate downloadable package for macOS when I release SLiM 3.4, so Mac users shouldn't need to build QtSLiM themselves, in general.  So that's great; that will make it much more accessible, since many Mac users (like me!) aren't very comfortable with doing builds at the command line.

Speaking of making it easier for users to install, @brnorris03  wrote:

> It's nice to have a Debian package (works on many distros), but of course, someone has to maintain it, i.e., update for new distro releases in addition to the SLiM updates. Those are not that frequent, so perhaps this is fairly easy.

Maybe it would be a good idea to do this.  @brnorris03, do you know how, and do you have time to put it together?  Graham Gower has put up an Arch package; you can see the source for it at https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=slim-simulator.  If a Debian one would be similarly simple, that would be nice to provide, it seems to me.

**Comment by petrelharp** [comment]
Perhaps @jgallowa07 could have a go, for a fresh set of eyes?

**Comment by petrelharp** [comment]
Small note about the docs: it should say "code from tskit", not msprime.

**Comment by petrelharp** [comment]
Other notes:
- you've still got instructions on installing cmake in macOS in the Linux section (although arguably, macOS falls under the "other Un*x platforms" heading?)  Also, you should probably not instruct people to download cmake directly, but to install via their package manager; so I'd change that sentence to "On Linux and other Un*x systems, you can install cmake using your package manager (e.g., `sudo apt-get install cmake` on Debian), or directly from the cmake home page at https://cmake.org/download."  (and, btw, I don't think the apologies are in order: after all, you don't apologize that people need to install a compiler to compile it...)
- I'd delete the phrase "If you're a Linux user you probably know all this already." - no need to push the idea that you have to know all this stuff to be a linux user. I didn't know all this for the majority of my linux-using years.
- It should say "On Debian *stable* it currently yields Qt 5.11.3"
- you've got the stuff on finding qmake in two separate places

**Comment by bhaller** [comment]
> * you've still got instructions on installing cmake in macOS in the Linux section (although arguably, macOS falls under the "other Un_x platforms" heading?)

Yeah, that's the idea; the previous section about building on macOS uses the Xcode IDE, so some Mac users will end up following the Unix instructions instead.  I've improved the first paragraph of that section to clarify that macOS is a Un*x platform.

> * I'd delete the phrase "If you're a Linux user you probably know all this already." - no need to push the idea that you have to know all this stuff to be a linux user. I didn't know all this for the majority of my linux-using years.

OK

> * It should say "On Debian _stable_ it currently yields Qt 5.11.3"

Ah, I finally get it – you make a choice for your whole Debian installation as to whether you're "stable" or "unstable" or "experimental", and get different versions of *everything* accordingly.  Is that right?  I thought it was an option you would pass to apt-get.  Or am I still confused?

> * you've got the stuff on finding qmake in two separate places

Do I?  Can you point me to where?

**Comment by petrelharp** [comment]
> you make a choice for your whole Debian installation as to whether you're "stable" or "unstable" or "experimental", and get different versions of everything accordingly

Yes! Like with Ubuntu, you might currently be `focal` or `bionic`; and with `focal` you'd get more up-to-date things...

> point me to where

`find /usr -name qmake -print` etc appears on both p.54 and p.58.

**Comment by jgallowa07** [comment]
> Perhaps @jgallowa07 could have a go, for a fresh set of eyes?

Happy to give this a try sometime this week when I find some free time. I'm running ubuntu 18 as VM so we'll see what happens :)

**Comment by petrelharp** [comment]
Rad. It should be quick, just a few `apt-get install`s and `make`s....

**Comment by bhaller** [comment]
OK, I just posted a new draft of the manual incorporating both @petrelharp's latest comments (no need to look at it again, though, Peter) and various changes prompted by me spending my afternoon building QtSLiM in a half a dozen different ways.  :->  I am hoping this is now final.  I plan to ship SLiM 3.4 tomorrow morning unless I hear about new issues before then.  Thanks everyone!

**Comment by bhaller** [comment]
OK, SLiM 3.4 is tagged in GitHub, although I haven't announced it publicly yet (I plan to do so tomorrow morning).  @grahamgower, if you could update your Arch package that would be great.  Thanks!

I'm going to close this issue now, since it was about how to roll QtSLiM in SLiM 3.4.  If new things arise, please open new issues.  :->

**Comment by grahamgower** [comment]
> OK, SLiM 3.4 is tagged in GitHub, although I haven't announced it publicly yet (I plan to do so tomorrow morning). @grahamgower, if you could update your Arch package that would be great. Thanks!

Updated! Thanks @bhaller!
